File: 20-ValidParentheses/20-ValidParentheses.py
class Solution(object):
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        # Initiate a stack of characters
        stack = []
        closeToOpen = { ")" : "(", "]" : "[", "}" : "{"} # hashmap of closing brackets to open

            
        for char in s:
            if char in closeToOpen:
                if stack and stack[-1] == closeToOpen[char]: # if closing bracket match open, pop them
                    stack.pop()
                else:
                    return False
            else:
                stack.append(char) # append opening bracket to stack
        
        return True if not stack else False

                 

                
        
        
        

File: 20-ValidParentheses/20-ValidParentheses.java
import java.util.HashMap;
import java.util.ArrayList;

class Solution {
    public boolean isValid(String s) {
        char[] sArray = s.toCharArray();
        List<Character> stack = new ArrayList<>();
        HashMap<Character, Character> closeToOpen = new HashMap<>();
        closeToOpen.put(')', '(');
        closeToOpen.put('}', '{');
        closeToOpen.put(']', '[');        

        for (char c : sArray) {
            if (closeToOpen.containsKey(c)) {
                if (stack.size() != 0 && stack.get(stack.size() - 1) == closeToOpen.get(c)) {
                    stack.remove(stack.size() - 1);
                }
                else {
                    return false;
                }
            } else {
                stack.add(c);
            }
        }

        if (stack.size() != 0) {
            return false;
        }

        return true;
    }
}